### ISSUE TITLE
Conditional @auth(orization)

### ISSUE THREAD

**Issue by bloomdido** [neo4j/graphql#163]
I'd like to be able to create custom authorization checks utilizing the `@cypher` directive. An example of this would be to check if one user is connected to another user via a matched path (that can't be matched via a `@relationship`):

```graphql
type Group {
  id: ID! @id
  name: String!
  members: [User!] @relationship(type: ":MEMBER_OF", direction: IN)
}

type User {
  id: ID! @id
  name: String!
  groups: [Group!] @relationship(type: ":MEMBER_OF", direction: OUT)

  isConnected: Boolean @cypher(statement: """
    MATCH (this)-[:MEMBER_OF]->(:Group)<-[:MEMBER_OF]-(u:User)
      WHERE u.id = $auth.jwt.sub
    RETURN count(u) > 0
  """)
}

extend type User {
  @auth(rules: [{
    operations: [READ]
    allow: {isConnected: true}
  }])
}
```

This would allow the current user to view another user _only_ if they were connected via a shared `Group`.

I can think of many other situations where this kind of flexibility would be useful and prevent the need for creating custom resolvers. It might even solve #159.

**Comment by bloomdido** [comment]
I also think this would solve the problem of more complicated roles-based authorization beyond the scope of JWT roles...

**Comment by danstarns** [comment]
Hey @bloomdido thanks for the suggestion, I do see a need for allowing custom auth Quieres backed by arbitrary Cypher, although looking at your example here -  It can be replicated using our supported feature [`allow`](https://neo4j.com/docs/graphql-manual/1.0-preview/auth/authorization/allow/): 

```graphql
type Group {
  id: ID! @id
  name: String!
  members: [User!] @relationship(type: ":MEMBER_OF", direction: IN)
}

type User {
  id: ID! @id
  name: String!
  groups: [Group!] @relationship(type: ":MEMBER_OF", direction: OUT)
}

extend type User {
  @auth(rules: [{
    operations: [READ]
    allow: {
      groups: {
        members: {
          id: "$jwt.sub"
        }
      }
    }
  }])
}
```
I would be really keen to see if you have a more complicated use case that would require arbitrary Cypher.

**Comment by bloomdido** [comment]
@danstarns Thanks! It's not clear from the documentation that nested relationships can be used like that in the `@auth allow` rules. I was able to get it to work for the above instance, as well as a few others.

I do think there's still value in using `@cypher` statements for conditional `@auth` rules. For one, it keeps the logic more centralized. Changes to schema could require significant rewrites of a lot of `@auth` logic scattered throughout. Second, it would provide a way of checking roles at query time, rather than when a JWT is generated. This could make an app more secure. Lastly, I think I still have some `@auth` cases that use more complex graph patterns than are currently feasible with `@relationship` directives. When I come across one of those cases (probably over the next few days), I'll add it here.

**Comment by realL4mbd4** [comment]
I have a similar issue @danstarns 

trying to make this pattern work:

`(user:User)<-[:SHARED_WITH]-(collection:Collection)<-[:INSIDE_OF*]-(post:Post)`
and
`(user:User)<-[:SHARED_WITH]-(collection:Collection)<-[:INSIDE_OF*]-(nestedCollection:Collection)`

Users should be able to access posts and collections that are inside of a collection shared with them or inside a collection that is a child of a collection of any depth that is shared with them.

Here's how I'm trying to work with this pattern.
`  type Collection @auth(rules: [
    {
      allow: {
        OR: [
          { team: { members: { _id: "$jwt.sub" } } },
          { guests: { _id: "$jwt.sub" } },
          { isGuest: true }
        ]
      }
    }
  ]) {
    _id: ID! @id
    name: String

    path: [Collection] @cypher(
      statement: """
      MATCH (this)-[:INSIDE_OF*]->(nodes:Collection)
      RETURN nodes
      """
    ) 
    workspace: Workspace @relationship(type: "WORKSPACE", direction: OUT)
    team: Team @relationship(type: "TEAM", direction: OUT)

    isGuest: Boolean @cypher(
      statement: """
      MATCH (user:User { _id: "$auth.jwt.sub" })
      WITH (
        EXISTS((this)-[:SHARED_WITH]->(user)) OR
        EXISTS((this)-[:INSIDE_OF*]->(:Collection)-[:SHARED_WITH]->(user))
      ) AS isGuest
      RETURN isGuest
      """
    )

    guests: [User] @relationship(type: "SHARED_WITH", direction: OUT)
    collections: [Collection] @relationship(type: "INSIDE_OF", direction: IN)
    posts: [Post] @relationship(type: "INSIDE_OF", direction: IN)
  }`

When I try to use this I get an error `value.split is not a function or its return value is not iterable`

Is there a supported allow pattern I could use to make something like this work?

Excited to start using neo4j/graphql in production, thank you for all your work on this project!

**Comment by bloomdido** [comment]
The `value.split` error comes about because of this:

https://github.com/neo4j/graphql/blob/6fdaedabe4fd5935ba8dc53a5029a8742de6e61c/packages/graphql/src/translate/create-auth-and-params.ts#L94-L96

It looks like if there was some logic to allow the potential for a boolean value, it would work for straight node properties. However, it would still fail for custom `@cypher` directives because of:

https://github.com/neo4j/graphql/blob/6fdaedabe4fd5935ba8dc53a5029a8742de6e61c/packages/graphql/src/translate/create-auth-and-params.ts#L110

I'd really like to see this feature implement simply because I am not a big fan of putting roles in a JWT, and this would allow for things like `isAdmin` properties to be checked at the time of access using custom cypher rather than having to calculate it only when the JWT is issued.

**Comment by jacob-buscher** [comment]
Is there a way to do namespace based [RBAC](https://neo4j.com/docs/graphql-manual/current/auth/authorization/roles/#_rbac) roles? @bloomdido @danstarns 

Something like `${teamId}:admin` or `${teamId}:owner` where `teamId` is populated from based on the query?

I'm having a hard time using the `@auth` directive to make tenant based roles and permissions.

**Comment by jacob-buscher** [comment]
@danstarns 

Do you know if this is going to be added or should we start working on a work around?

Thank you for your help on this!

**Comment by darrellwarde** [comment]
Hey @jacob-buscher, we're sorry this doesn't offer the features you need yet! Honestly, our hands are going to be really full for at least the next few weeks as we build up to the 2.0.0 with relationship properties, so we're not going to have time to take a look at this. 

So if you can find a workaround, please do. Alternatively, we're always happy to accept PRs for work we just don't have time do, we can always find time for a review! If you do fancy it, we would like the opportunity to discuss the solution a bit more first. Let us know! 🙂

**Comment by jacob-buscher** [comment]
@darrellwarde

Love version 2.0.0, will definitely accelerate our dev!

Do you know if anyone is working on conditional auth like this?

**Comment by darrellwarde** [comment]
Hey @jacob-buscher, thanks for the kind words! 🙂 

I'm afraid no one is looking at this right now, and our focus at the moment is on much more substantial, overarching features. Apologies!

**Comment by darrellwarde** [comment]
We're thinking about our long-term roadmap and want to consolidate our backlog so that we can accurately plan moving forward. As such, this is being closed in favour of #1871 where we will capture all requirements in relation to filtering in the `@auth` directive.
